Extracted Pulsar Function jar file content in /tmp/pulsar-nar directory gets corrupted or overridden by different content

**Describe the bug**

By default, Pulsar Functions runtime extracts files of a Pulsar Function implementation jar file to a directory `/tmp/pulsar-nar/[function-jar-file-name].jar-unpacked` where the "function-jar-file-name" part is determined used a logic which doesn't produce unique file name.

When the Pulsar Function is installed from a URL source over http/https, the filename seems to always be the short name of the package for the implementation class. 
For example for a function class `com.github.lhotari.pulsar.RandomFilterFunction`, the filename becomes `pulsar.jar`. This filename would be the same for all function classes in the same package, regardless of the implementation classname or it's version.

The logic is here:
https://github.com/apache/pulsar/blob/3fcfb22e71c9774f710d659030d4af01eb05c956/pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionActioner.java#L548-L572

The logic for extracting the Nar file is here:
https://github.com/apache/pulsar/blob/c2e5ec6a5d4b4f21df37d1f2c9b878cf801d6f1a/pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarUnpacker.java#L60-L85

As it can be seen from this code, it leads to collisions when there are multiple jar files with the same name.
Another problem can happen when 2 threads are in a race in extracting the nar file.
It can lead to an error message such as this one in the Pulsar Functions instance log file:

```
07:20:38.336 [public/default/random-1-2] INFO  org.apache.pulsar.functions.instance.JavaInstanceRunnable - Load JAR: /tmp/pulsar_functions/public/default/random-1/2/pulsar.jar
07:20:38.378 [public/default/random-1-2] WARN  org.apache.pulsar.common.nar.NarClassLoader - /tmp/pulsar-nar/pulsar.jar-unpacked does not contain META-INF/bundled-dependencies!
07:20:38.378 [public/default/random-1-2] INFO  org.apache.pulsar.functions.instance.JavaInstanceRunnable - Initialize function class loader for function random-1 at function cache manager, functionClassLoader: org.apache.pulsar.common.nar.NarClassLoader[/tmp/pulsar-nar/pulsar.jar-unpacked]
07:20:38.381 [public/default/random-1-2] ERROR org.apache.pulsar.functions.instance.JavaInstanceRunnable - [public/default/random-1:2] Uncaught exception in Java Instance
java.lang.RuntimeException: User class must be in class path
        at org.apache.pulsar.common.util.Reflections.createInstance(Reflections.java:112) ~[java-instance.jar:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setup(JavaInstanceRunnable.java:197) ~[org.apache.pulsar-pulsar-functions-instance-2.7.2.1.1.0.jar:2.7.2.1.1.0]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run(JavaInstanceRunnable.java:235) ~[org.apache.pulsar-pulsar-functions-instance-2.7.2.1.1.0.jar:2.7.2.1.1.0]
        at java.lang.Thread.run(Thread.java:829) ~[?:?]
Caused by: java.lang.ClassNotFoundException: com.github.lhotari.pulsar.RandomFilterFunction
        at java.net.URLClassLoader.findClass(URLClassLoader.java:471) ~[?:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:589) ~[?:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:522) ~[?:?]
        at java.lang.Class.forName0(Native Method) ~[?:?]
        at java.lang.Class.forName(Class.java:398) ~[?:?]
        at org.apache.pulsar.common.util.Reflections.createInstance(Reflections.java:110) ~[java-instance.jar:?]
        ... 3 more
07:20:38.390 [public/default/random-1-2] INFO  org.apache.pulsar.functions.instance.JavaInstanceRunnable - Closing instance
```


**To Reproduce**
Steps to reproduce the behavior:

Install a function with process runtime with parallelism > brokers/function workers. This can reproduce the race condition when /tmp/pulsar-nar doesn't get include the `[function-jar-file-name].jar-unpacked` directory. It might require a few attempts to reproduce the issue.

For example:
```
pulsar-admin functions create --tenant public --namespace default --name random-1 --parallelism 8 --inputs persistent://public/default/random-1-in --output persistent://amplitude/processing/random-2-in --jar https://github.com/lhotari/pulsar-issue-6054-repro/releases/download/v1.0/random-filter-function-1.0-SNAPSHOT.jar --ram 209715200 --classname com.github.lhotari.pulsar.RandomFilterFunction
```

**Expected behavior**

Pulsar Functions loading shouldn't mix different function implementations by extracting them to the same directory. Race conditions should also be handled when the same function is started with multiple instances on the same node.

--- a/pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarUnpacker.java
+++ b/pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarUnpacker.java
@@ -75,9 +75,9 @@ public class NarUnpacker {
                 throw new IOException("Cannot create " + parentDirectory);
             }
         }
-        String sha256Sum = Base64.getUrlEncoder().withoutPadding().encodeToString(calculateSha256Sum(nar));
+        String md5Sum = Base64.getUrlEncoder().withoutPadding().encodeToString(calculateMd5sum(nar));
         // ensure that one process can extract the files
-        File lockFile = new File(parentDirectory, "." + sha256Sum + ".lock");
+        File lockFile = new File(parentDirectory, "." + md5Sum + ".lock");
         // prevent OverlappingFileLockException by ensuring that one thread tries to create a lock in this JVM
         Object localLock = CURRENT_JVM_FILE_LOCKS.computeIfAbsent(lockFile.getAbsolutePath(), key -> new Object());
         synchronized (localLock) {
@@ -85,7 +85,7 @@ public class NarUnpacker {
             // using the same lock file don't execute concurrently
             try (FileChannel channel = new RandomAccessFile(lockFile, "rw").getChannel();
                  FileLock lock = channel.lock()) {
-                File narWorkingDirectory = new File(parentDirectory, sha256Sum);
+                File narWorkingDirectory = new File(parentDirectory, md5Sum);
                 if (narWorkingDirectory.mkdir()) {
                     try {
                         log.info("Extracting {} to {}", nar, narWorkingDirectory);
@@ -150,27 +150,27 @@ public class NarUnpacker {
     }
 
     /**
-     * Calculates an sha256 sum of the specified file.
+     * Calculates an md5 sum of the specified file.
      *
      * @param file
-     *            to calculate the sha256 of
-     * @return the sha256 bytes
+     *            to calculate the md5sum of
+     * @return the md5sum bytes
      * @throws IOException
      *             if cannot read file
      */
-    private static byte[] calculateSha256Sum(final File file) throws IOException {
+    private static byte[] calculateMd5sum(final File file) throws IOException {
         try (final FileInputStream inputStream = new FileInputStream(file)) {
-            final MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
+            final MessageDigest md5 = MessageDigest.getInstance("md5");
 
             final byte[] buffer = new byte[1024];
             int read = inputStream.read(buffer);
 
             while (read > -1) {
-                sha256.update(buffer, 0, read);
+                md5.update(buffer, 0, read);
                 read = inputStream.read(buffer);
             }
 
-            return sha256.digest();
+            return md5.digest();
         } catch (NoSuchAlgorithmException nsae) {
             throw new IllegalArgumentException(nsae);
         }

--- a/pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarUnpacker.java
+++ b/pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarUnpacker.java
@@ -24,16 +24,20 @@
 
 package org.apache.pulsar.common.nar;
 
+import com.google.common.annotations.VisibleForTesting;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.nio.file.Files;
+import java.io.RandomAccessFile;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
-import java.util.Arrays;
+import java.util.Base64;
 import java.util.Enumeration;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
 import lombok.extern.slf4j.Slf4j;
@@ -43,8 +47,7 @@ import lombok.extern.slf4j.Slf4j;
  */
 @Slf4j
 public class NarUnpacker {
-
-    private static final String HASH_FILENAME = "nar-md5sum";
+    private static final ConcurrentHashMap<String, Object> CURRENT_JVM_FILE_LOCKS = new ConcurrentHashMap<>();
 
     /**
      * Unpacks the specified nar into the specified base working directory.
@@ -58,43 +61,55 @@ public class NarUnpacker {
      *             if unable to explode nar
      */
     public static File unpackNar(final File nar, final File baseWorkingDirectory) throws IOException {
-        final File narWorkingDirectory = new File(baseWorkingDirectory, nar.getName() + "-unpacked");
+        return doUnpackNar(nar, baseWorkingDirectory, null);
+    }
 
-        // if the working directory doesn't exist, unpack the nar
-        if (!narWorkingDirectory.exists()) {
-            unpack(nar, narWorkingDirectory, calculateMd5sum(nar));
-        } else {
-            // the working directory does exist. Run MD5 sum against the nar
-            // file and check if the nar has changed since it was deployed.
-            final byte[] narMd5 = calculateMd5sum(nar);
-            final File workingHashFile = new File(narWorkingDirectory, HASH_FILENAME);
-            if (!workingHashFile.exists()) {
-                FileUtils.deleteFile(narWorkingDirectory, true);
-                unpack(nar, narWorkingDirectory, narMd5);
-            } else {
-                final byte[] hashFileContents = Files.readAllBytes(workingHashFile.toPath());
-                if (!Arrays.equals(hashFileContents, narMd5)) {
-                    log.info("Contents of nar {} have changed. Reloading.", nar.getAbsolutePath());
-                    FileUtils.deleteFile(narWorkingDirectory, true);
-                    unpack(nar, narWorkingDirectory, narMd5);
+    @VisibleForTesting
+    static File doUnpackNar(final File nar, final File baseWorkingDirectory, Runnable extractCallback)
+            throws IOException {
+        File parentDirectory = new File(baseWorkingDirectory, nar.getName() + "-unpacked");
+        if (!parentDirectory.exists()) {
+            parentDirectory.mkdirs();
+        }
+        String sha256Sum = Base64.getUrlEncoder().withoutPadding().encodeToString(calculateSha256Sum(nar));
+        // ensure that one process can extract the files
+        File lockFile = new File(parentDirectory, "." + sha256Sum + ".lock");
+        // prevent OverlappingFileLockException by ensuring that one thread tries to create a lock in this JVM
+        Object localLock = CURRENT_JVM_FILE_LOCKS.computeIfAbsent(lockFile.getAbsolutePath(), key -> new Object());
+        synchronized (localLock) {
+            // create file lock that ensures that other processes
+            // using the same lock file don't execute concurrently
+            try (FileChannel channel = new RandomAccessFile(lockFile, "rw").getChannel();
+                 FileLock lock = channel.lock()) {
+                File narWorkingDirectory = new File(parentDirectory, sha256Sum);
+                if (narWorkingDirectory.mkdir()) {
+                    try {
+                        log.info("Extracting {} to {}", nar, narWorkingDirectory);
+                        if (extractCallback != null) {
+                            extractCallback.run();
+                        }
+                        unpack(nar, narWorkingDirectory);
+                    } catch (IOException e) {
+                        log.error("There was a problem extracting the nar file. Deleting {} to clean up state.",
+                                narWorkingDirectory, e);
+                        FileUtils.deleteFile(narWorkingDirectory, true);
+                        throw e;
+                    }
                 }
+                return narWorkingDirectory;
             }
         }
-
-        return narWorkingDirectory;
     }
 
     /**
-     * Unpacks the NAR to the specified directory. Creates a checksum file that used to determine if future expansion is
-     * necessary.
+     * Unpacks the NAR to the specified directory.
      *
      * @param workingDirectory
      *            the root directory to which the NAR should be unpacked.
      * @throws IOException
      *             if the NAR could not be unpacked.
      */
-    private static void unpack(final File nar, final File workingDirectory, final byte[] hash) throws IOException {
-
+    private static void unpack(final File nar, final File workingDirectory) throws IOException {
         try (JarFile jarFile = new JarFile(nar)) {
             Enumeration<JarEntry> jarEntries = jarFile.entries();
             while (jarEntries.hasMoreElements()) {
@@ -107,11 +122,6 @@ public class NarUnpacker {
                     makeFile(jarFile.getInputStream(jarEntry), f);
                 }
             }
-        }
-
-        final File hashFile = new File(workingDirectory, HASH_FILENAME);
-        try (final FileOutputStream fos = new FileOutputStream(hashFile)) {
-            fos.write(hash);
         }
     }
 
@@ -136,27 +146,27 @@ public class NarUnpacker {
     }
 
     /**
-     * Calculates an md5 sum of the specified file.
+     * Calculates an sha256 sum of the specified file.
      *
      * @param file
-     *            to calculate the md5sum of
-     * @return the md5sum bytes
+     *            to calculate the sha256 of
+     * @return the sha256 bytes
      * @throws IOException
      *             if cannot read file
      */
-    private static byte[] calculateMd5sum(final File file) throws IOException {
+    private static byte[] calculateSha256Sum(final File file) throws IOException {
         try (final FileInputStream inputStream = new FileInputStream(file)) {
-            final MessageDigest md5 = MessageDigest.getInstance("md5");
+            final MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
 
             final byte[] buffer = new byte[1024];
             int read = inputStream.read(buffer);
 
             while (read > -1) {
-                md5.update(buffer, 0, read);
+                sha256.update(buffer, 0, read);
                 read = inputStream.read(buffer);
             }
 
-            return md5.digest();
+            return sha256.digest();
         } catch (NoSuchAlgorithmException nsae) {
             throw new IllegalArgumentException(nsae);
         }

--- a/pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarUnpacker.java
+++ b/pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarUnpacker.java
@@ -69,7 +69,11 @@ public class NarUnpacker {
             throws IOException {
         File parentDirectory = new File(baseWorkingDirectory, nar.getName() + "-unpacked");
         if (!parentDirectory.exists()) {
-            parentDirectory.mkdirs();
+            if (parentDirectory.mkdirs()) {
+                log.info("Created directory {}", parentDirectory);
+            } else if (!parentDirectory.exists()) {
+                throw new IOException("Cannot create " + parentDirectory);
+            }
         }
         String sha256Sum = Base64.getUrlEncoder().withoutPadding().encodeToString(calculateSha256Sum(nar));
         // ensure that one process can extract the files

--- a/pulsar-common/src/test/java/org/apache/pulsar/common/nar/NarUnpackerTest.java
+++ b/pulsar-common/src/test/java/org/apache/pulsar/common/nar/NarUnpackerTest.java
@@ -1,0 +1,160 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.nar;
+
+import static org.junit.Assert.assertTrue;
+import static org.testng.Assert.assertEquals;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.SystemUtils;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+@Slf4j
+public class NarUnpackerTest {
+    File sampleZipFile;
+    File extractDirectory;
+
+    @BeforeMethod
+    public void createSampleZipFile() throws IOException {
+        sampleZipFile = Files.createTempFile("sample", ".zip").toFile();
+        try (ZipOutputStream out = new ZipOutputStream(new FileOutputStream(sampleZipFile))) {
+            for (int i = 0; i < 10000; i++) {
+                ZipEntry e = new ZipEntry("hello" + i + ".txt");
+                out.putNextEntry(e);
+                byte[] msg = "hello world!".getBytes(StandardCharsets.UTF_8);
+                out.write(msg, 0, msg.length);
+                out.closeEntry();
+            }
+        }
+        extractDirectory = Files.createTempDirectory("nar_unpack_dir").toFile();
+    }
+
+    @AfterMethod(alwaysRun = true)
+    void deleteSampleZipFile() throws IOException {
+        if (sampleZipFile != null) {
+            sampleZipFile.delete();
+        }
+        if (extractDirectory != null) {
+            FileUtils.deleteFile(extractDirectory, true);
+        }
+    }
+
+    @Test
+    void shouldExtractFilesOnceInSameProcess() throws InterruptedException {
+        int threads = 20;
+        CountDownLatch countDownLatch = new CountDownLatch(threads);
+        AtomicInteger exceptionCounter = new AtomicInteger();
+        AtomicInteger extractCounter = new AtomicInteger();
+        for (int i = 0; i < threads; i++) {
+            new Thread(() -> {
+                try {
+                    NarUnpacker.doUnpackNar(sampleZipFile, extractDirectory, extractCounter::incrementAndGet);
+                } catch (Exception e) {
+                    e.printStackTrace();
+                    exceptionCounter.incrementAndGet();
+                } finally {
+                    countDownLatch.countDown();
+                }
+            }).start();
+        }
+        assertTrue(countDownLatch.await(30, TimeUnit.SECONDS));
+        assertEquals(exceptionCounter.get(), 0);
+        assertEquals(extractCounter.get(), 1);
+    }
+
+    public static class NarUnpackerWorker {
+        public static void main(String[] args) {
+            File sampleZipFile = new File(args[0]);
+            File extractDirectory = new File(args[1]);
+            AtomicInteger extractCounter = new AtomicInteger();
+            try {
+                NarUnpacker.doUnpackNar(sampleZipFile, extractDirectory, extractCounter::incrementAndGet);
+                if (extractCounter.get() == 1) {
+                    System.exit(101);
+                } else if (extractCounter.get() == 0) {
+                    System.exit(100);
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+                System.exit(99);
+            }
+        }
+    }
+
+    @Test
+    void shouldExtractFilesOnceInDifferentProcess() throws InterruptedException {
+        int processes = 10;
+        String javaExePath = findJavaExe().getAbsolutePath();
+        CountDownLatch countDownLatch = new CountDownLatch(processes);
+        AtomicInteger exceptionCounter = new AtomicInteger();
+        AtomicInteger extractCounter = new AtomicInteger();
+        for (int i = 0; i < processes; i++) {
+            new Thread(() -> {
+                try {
+                    // fork a new process with the same classpath
+                    Process process = new ProcessBuilder()
+                            .command(javaExePath,
+                                    "-Xmx64m",
+                                    "-cp",
+                                    System.getProperty("java.class.path"),
+                                    // use NarUnpackerWorker as the main class
+                                    NarUnpackerWorker.class.getName(),
+                                    // pass arguments to use for testing
+                                    sampleZipFile.getAbsolutePath(),
+                                    extractDirectory.getAbsolutePath())
+                            .start();
+                    String output = IOUtils.toString(process.getInputStream(), StandardCharsets.UTF_8);
+                    int retval = process.waitFor();
+                    log.info("Process retval {} output {}", retval, output);
+                    if (retval == 101) {
+                        extractCounter.incrementAndGet();
+                    } else if (retval != 100) {
+                        exceptionCounter.incrementAndGet();
+                    }
+                } catch (Exception e) {
+                    e.printStackTrace();
+                    exceptionCounter.incrementAndGet();
+                } finally {
+                    countDownLatch.countDown();
+                }
+            }).start();
+        }
+        assertTrue(countDownLatch.await(30, TimeUnit.SECONDS));
+        assertEquals(exceptionCounter.get(), 0);
+        assertEquals(extractCounter.get(), 1);
+    }
+
+    File findJavaExe() {
+        File javaHome = new File(System.getProperty("java.home"));
+        File javaExe = new File(javaHome, "bin/java" + (SystemUtils.IS_OS_WINDOWS ? ".exe" : ""));
+        return javaExe;
+    }
+}

--- a/pulsar-common/src/test/java/org/apache/pulsar/common/nar/NarUnpackerTest.java
+++ b/pulsar-common/src/test/java/org/apache/pulsar/common/nar/NarUnpackerTest.java
@@ -78,7 +78,7 @@ public class NarUnpackerTest {
                 try {
                     NarUnpacker.doUnpackNar(sampleZipFile, extractDirectory, extractCounter::incrementAndGet);
                 } catch (Exception e) {
-                    e.printStackTrace();
+                    log.error("Unpacking failed", e);
                     exceptionCounter.incrementAndGet();
                 } finally {
                     countDownLatch.countDown();
@@ -103,7 +103,7 @@ public class NarUnpackerTest {
                     System.exit(100);
                 }
             } catch (Exception e) {
-                e.printStackTrace();
+                log.error("Unpacking failed", e);
                 System.exit(99);
             }
         }
@@ -140,7 +140,7 @@ public class NarUnpackerTest {
                         exceptionCounter.incrementAndGet();
                     }
                 } catch (Exception e) {
-                    e.printStackTrace();
+                    log.error("Unpacking in a separate process failed", e);
                     exceptionCounter.incrementAndGet();
                 } finally {
                     countDownLatch.countDown();

